File: 2022/july/24-07-2022/search-a-2d-matrix-ii.py
class Solution:
    def searchMatrix(self, matrix, target: int) -> bool:
        row, col = len(matrix)-1, 0

        while row >=0 and col <= len(matrix[0])-1:
            cell = matrix[row][col]
            if cell > target:
                row-= 1
            elif cell < target:
                col+= 1
            else: return True

        return False


print(Solution().searchMatrix([[1,4,7,11,15],[2,5,8,12,19],[3,6,9,16,22],[10,13,14,17,24],[18,21,23,26,30]], 25))

File: 2022/july/24-07-2022/search-a-2d-matrix-ii.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let (mut i, mut j) = (0, matrix[0].len() - 1);
        while i < matrix.len() {
            match matrix[i][j].cmp(&target) {
                std::cmp::Ordering::Less => i += 1,
                std::cmp::Ordering::Equal => return true,
                std::cmp::Ordering::Greater if j > 0 => j -= 1,
                std::cmp::Ordering::Greater => break,
            }
        }
        false
    }
}